src/Serialization.cpp

```diff
@@ -0,0 +1,1310 @@
+    int32_t dimensions = buffer.dimensions();
+    std::vector<flatbuffers::Offset<Halide::Serialize::BufferDimension>> buffer_dimensions_serialized;
+    for (int i = 0; i < buffer.dimensions(); ++i) {
+        int32_t min = buffer.dim(i).min();
+        int32_t extent = buffer.dim(i).extent();
+        int32_t stride = buffer.dim(i).stride();
+        buffer_dimensions_serialized.push_back(Halide::Serialize::CreateBufferDimension(builder, min, extent, stride));
+    }
+    std::vector<uint8_t> data;
+    data.resize(buffer.size_in_bytes());
+    memcpy(data.data(), buffer.data(), buffer.size_in_bytes());
+    return Halide::Serialize::CreateBuffer(builder, true, name_serialized, type_serialized, dimensions, builder.CreateVector(buffer_dimensions_serialized), builder.CreateVector(data));
+}
+
+std::vector<flatbuffers::Offset<Halide::Serialize::WrapperRef>> Serializer::serialize_wrapper_refs(flatbuffers::FlatBufferBuilder &builder, const std::map<std::string, FunctionPtr> &wrappers) {
+    std::vector<flatbuffers::Offset<Halide::Serialize::WrapperRef>> wrapper_refs_serialized;
+    wrapper_refs_serialized.reserve(wrappers.size());
+    for (const auto &wrapper : wrappers) {
+        auto wrapper_name_serialized = serialize_string(builder, wrapper.first);
+        int func_index = -1;
+        if (this->func_mappings.find(Function(wrapper.second).name()) != this->func_mappings.end()) {
+            func_index = this->func_mappings[Function(wrapper.second).name()];
+        }
+        wrapper_refs_serialized.push_back(Halide::Serialize::CreateWrapperRef(builder, wrapper_name_serialized, func_index));
+    }
+    return wrapper_refs_serialized;
+}
+
+void Serializer::build_function_mappings(const std::map<std::string, Function> &env) {
+    if (!this->func_mappings.empty()) {
+        this->func_mappings.clear();
+    }
+    int32_t cnt = 0;
+    for (const auto &entry : env) {
+        user_assert(env.find(entry.first) != env.end()) << "function " << entry.first << " not found in the environment\n";
+        this->func_mappings[entry.first] = cnt++;
+    }
+}
+
+void Serializer::serialize(const Pipeline &pipeline, const std::string &filename) {
+    flatbuffers::FlatBufferBuilder builder(1024);
+
+    // extract the DAG, unwarp function from Funcs
+    std::vector<Function> outputs_functions;
+    for (const Func &func : pipeline.outputs()) {
+        outputs_functions.push_back(func.function());
+    }
+    std::map<std::string, Function> env = build_environment(outputs_functions);
+    build_function_mappings(env);
+
+    std::vector<flatbuffers::Offset<flatbuffers::String>> func_names_in_order_serialized;
+    std::vector<flatbuffers::Offset<Halide::Serialize::Func>> funcs_serialized;
+    for (const auto &entry : env) {
+        func_names_in_order_serialized.push_back(serialize_string(builder, entry.first));
+        funcs_serialized.push_back(this->serialize_function(builder, entry.second));
+    }
+
+    auto outpus = pipeline.outputs();
```

Comment:
...`outputs`? (Or did you mean `octopus`?)

Reply:
I must be hungry when i wrote this line